File: native-linux/src/background/background_content_builder.cpp
#include "background_content_builder.h"
#include "app_container.h"
#include "data/db/event_dao.h"
#include "data/db/task_dao.h"
#include "data/db/issue_dao.h"
#include "data/prefs/prefs.h"
#include <algorithm>
#include <chrono>
#include <cctype>
#include <ranges>
#include <regex>

namespace cd {
namespace {
EpochMillis now_ms() { using namespace std::chrono; return duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count(); }
bool ci_equal(std::string a, std::string b) { for(char& c:a)c=static_cast<char>(std::tolower(static_cast<unsigned char>(c)));for(char& c:b)c=static_cast<char>(std::tolower(static_cast<unsigned char>(c)));return a==b; }
bool has_tag(std::vector<std::string> const& values, std::string const& tag) { return std::ranges::any_of(values,[&](auto const& v){return ci_equal(v,tag);}); }
bool date_matches(std::optional<EpochMillis> value, std::string const& filter) {
    if (filter == "all") return true;
    if (!value) return false;
    GDateTime* now=g_date_time_new_now_local(); GDateTime* today=g_date_time_new_local(g_date_time_get_year(now),g_date_time_get_month(now),g_date_time_get_day_of_month(now),0,0,0);
    GDateTime* tomorrow=g_date_time_add_days(today,1); GDateTime* after=g_date_time_add_days(today,2); GDateTime* week=g_date_time_add_days(today,-(g_date_time_get_day_of_week(today)-1)); GDateTime* next=g_date_time_add_days(week,7);
    auto ms=[](GDateTime* d){return EpochMillis(g_date_time_to_unix(d))*1000;};
    bool ok=filter=="today"?*value>=ms(today)&&*value<ms(tomorrow):filter=="tomorrow"?*value>=ms(tomorrow)&&*value<ms(after):*value>=ms(week)&&*value<ms(next);
    g_date_time_unref(next);g_date_time_unref(week);g_date_time_unref(after);g_date_time_unref(tomorrow);g_date_time_unref(today);g_date_time_unref(now);return ok;
}
std::string format_time(EpochMillis value) { GDateTime* d=g_date_time_new_from_unix_local(value/1000); gchar* s=g_date_time_format(d,"%b %e · %R"); std::string out=s?s:"";g_free(s);g_date_time_unref(d);return out; }
int estimate_minutes(std::string const& value) {
    int total{};
    try {
        std::regex const hours(R"((^|\s)#(\d+)\s*h\b)",std::regex::icase), minutes(R"((^|\s)#(\d+)\s*m\b)",std::regex::icase);
        for(std::sregex_iterator i(value.begin(),value.end(),hours),e;i!=e;++i)total+=std::stoi((*i)[2].str())*60;
        for(std::sregex_iterator i(value.begin(),value.end(),minutes),e;i!=e;++i)total+=std::stoi((*i)[2].str());
    } catch(...) {}
    return total;
}
}

BackgroundContent build_background_content(AppContainer& app, BackgroundTemplate const& t) {
    BackgroundContent out; out.title=t.source==BackgroundSource::Inbox?"INBOX":"VIEWS";
    out.filters=t.source==BackgroundSource::Inbox?t.inbox_type+" · "+t.inbox_date:t.views_type+" · "+t.views_date; out.mode=t.views_mode;
    auto const now=now_ms();
    if (t.source==BackgroundSource::Inbox) {
        auto const horizon=now+7LL*24*60*60*1000;
        if(t.inbox_type=="all"||t.inbox_type=="events") for(auto const& e:EventDao(app.db()).get_all()) if(e.start<=horizon&&std::max(e.end,e.start)>now&&date_matches(e.start,t.inbox_date)){out.rows.push_back({e.summary,format_time(e.start),"",0,false});out.total_minutes+=estimate_minutes(e.summary+" "+e.description.value_or(""));}
        if(t.inbox_type=="all"||t.inbox_type=="tasks") for(auto const& v:TaskDao(app.db()).get_all()) if(v.status!=TaskStatus::Completed&&date_matches(v.due,t.inbox_date)){out.rows.push_back({v.summary,v.due?format_time(*v.due):"No due date","",1,v.due&&*v.due<now});std::string tags=v.summary+" "+v.description.value_or("");for(auto const& tag:v.categories)tags+=" #"+tag;out.total_minutes+=estimate_minutes(tags);}
        if(t.inbox_type=="all"||t.inbox_type=="issues") for(auto const& v:IssueDao(app.db()).get_all()) if(v.state=="open"&&date_matches(v.milestone_due_on,t.inbox_date)){out.rows.push_back({v.title,v.repository,"",2,false});std::string tags=v.title+" "+v.body;for(auto const& tag:v.labels)tags+=" #"+tag;out.total_minutes+=estimate_minutes(tags);}
        std::stable_sort(out.rows.begin(),out.rows.end(),[](auto const& a,auto const& b){return a.title<b.title;});
    } else {
        auto settings=merged_app_preferences(app.prefs()); auto const& groups=settings.kanban_columns;
        if(t.views_type!="issues") for(auto const& v:TaskDao(app.db()).get_all()) if(v.status!=TaskStatus::Completed&&v.status!=TaskStatus::Cancelled&&date_matches(v.due,t.views_date)){
            std::string group;
            if(t.views_mode=="covey") { if(has_tag(v.categories,"do"))group="Do First";else if(has_tag(v.categories,"delay"))group="Schedule";else if(has_tag(v.categories,"delegate"))group="Delegate";else if(has_tag(v.categories,"eliminate"))group="Eliminate";else group="Untagged"; }
            else { auto i=std::ranges::find_if(groups,[&](auto const& g){return has_tag(v.categories,g);}); group=i==groups.end()?"Untagged":*i; }
            out.rows.push_back({v.summary,v.due?format_time(*v.due):"Task",group,1,v.due&&*v.due<now});
        }
        if(t.views_type!="tasks") for(auto const& v:IssueDao(app.db()).get_all()) if(v.state=="open"&&date_matches(v.milestone_due_on,t.views_date)){
            std::string group="Untagged";
            if(t.views_mode=="covey") { if(has_tag(v.labels,"do"))group="Do First";else if(has_tag(v.labels,"delay"))group="Schedule";else if(has_tag(v.labels,"delegate"))group="Delegate";else if(has_tag(v.labels,"eliminate"))group="Eliminate"; }
            else { auto i=std::ranges::find_if(groups,[&](auto const& g){return has_tag(v.labels,g);});if(i!=groups.end())group=*i; }
            if(t.views_mode!="covey"||group!="Untagged")out.rows.push_back({v.title,v.repository,group,2,false});
        }
    }
    return out;
}
}

File: native-android/app/src/main/kotlin/com/crossdashboard/app/background/BackgroundContentBuilder.kt
package com.crossdashboard.app.background

import com.crossdashboard.app.data.repository.EventRepository
import com.crossdashboard.app.data.repository.IssueRepository
import com.crossdashboard.app.data.repository.TaskRepository
import com.crossdashboard.app.data.prefs.AppPreferences
import com.crossdashboard.app.domain.model.TaskStatus
import com.crossdashboard.app.ui.screen.inbox.InboxViewModel
import kotlinx.coroutines.flow.first
import java.time.*
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BackgroundContentBuilder @Inject constructor(
    private val events: EventRepository,
    private val tasks: TaskRepository,
    private val issues: IssueRepository,
    private val prefs: AppPreferences,
) {
    suspend fun build(template: BackgroundTemplate): BackgroundContent = when (template.source) {
        BackgroundSource.INBOX -> buildInbox(template)
        BackgroundSource.VIEWS -> buildViews(template)
    }

    private suspend fun buildInbox(t: BackgroundTemplate): BackgroundContent {
        val now = Instant.now()
        data class TimedRow(val row: BackgroundRow, val date: Instant?, val minutes: Int)
        val rows = mutableListOf<TimedRow>()
        val eventValues = events.events.first()
        val taskValues = tasks.allTasks.first()
        val issueValues = issues.allIssues.first()
        if (t.inboxTypeFilter == "ALL" || t.inboxTypeFilter == "EVENTS") {
            eventValues.filter { it.end.isAfter(now.minus(Duration.ofDays(1))) }.sortedBy { it.start }.forEach { event ->
                val duration = Duration.between(event.start, event.end).toMinutes().toInt().takeIf { it in 1..<1440 } ?: 0
                rows += TimedRow(BackgroundRow(event.summary, formatDate(event.start), BackgroundRow.Kind.EVENT), event.start, duration)
            }
        }
        if (t.inboxTypeFilter == "ALL" || t.inboxTypeFilter == "TASKS") {
            taskValues.filter { it.status != TaskStatus.COMPLETED && it.status != TaskStatus.CANCELLED }
                .sortedBy { it.due }.forEach { task ->
                    rows += TimedRow(BackgroundRow(task.summary, task.due?.let(::formatDate) ?: "No due date", BackgroundRow.Kind.TASK,
                        overdue = task.due?.isBefore(now) == true), task.due, InboxViewModel.parseTimeEstimate(task.categories) ?: 0)
                }
        }
        if (t.inboxTypeFilter == "ALL" || t.inboxTypeFilter == "ISSUES") {
            issueValues.filter { it.state == "open" }.sortedByDescending { it.updatedAt }.forEach { issue ->
                rows += TimedRow(BackgroundRow(issue.title, issue.repository, BackgroundRow.Kind.ISSUE), issue.milestoneDueOn,
                    InboxViewModel.parseTimeEstimate(issue.labels) ?: 0)
            }
        }
        val filtered = rows.filter { matchesDate(it.date, t.inboxDateFilter) }
        return BackgroundContent("INBOX", "${label(t.inboxTypeFilter)} · ${label(t.inboxDateFilter)}",
            rows = filtered.map { it.row }, totalMinutes = filtered.sumOf { it.minutes })
    }

    private suspend fun buildViews(t: BackgroundTemplate): BackgroundContent {
        val columns = prefs.kanbanColumnsFlow.first()
        val rows = mutableListOf<BackgroundRow>()
        if (t.viewsTypeFilter != "ISSUES") {
            tasks.allTasks.first().filter { it.status != TaskStatus.COMPLETED && it.status != TaskStatus.CANCELLED && matchesDate(it.due, t.viewsDateFilter) }
                .forEach { task ->
                    val group = if (t.viewsMode == "COVEY") {
                        when {
                            "do" in task.categories -> "Do First"
                            "delay" in task.categories -> "Schedule"
                            "delegate" in task.categories -> "Delegate"
                            "eliminate" in task.categories -> "Eliminate"
                            else -> "Untagged"
                        }
                    } else task.categories.firstOrNull { it in columns } ?: "Untagged"
                    if (t.viewsMode != "COVEY" || group != "Untagged") {
                        rows += BackgroundRow(task.summary, task.due?.let(::formatDate) ?: "Task", BackgroundRow.Kind.TASK, group)
                    }
                }
        }
        if (t.viewsTypeFilter != "TASKS") {
            issues.allIssues.first().filter { it.state == "open" && matchesDate(it.milestoneDueOn, t.viewsDateFilter) }.forEach { issue ->
                val group = if (t.viewsMode == "COVEY") {
                    when {
                        "do" in issue.labels -> "Do First"
                        "delay" in issue.labels -> "Schedule"
                        "delegate" in issue.labels -> "Delegate"
                        "eliminate" in issue.labels -> "Eliminate"
                        else -> "Untagged"
                    }
                } else issue.labels.firstOrNull { it in columns } ?: "Untagged"
                if (t.viewsMode != "COVEY" || group != "Untagged") {
                    rows += BackgroundRow(issue.title, issue.repository, BackgroundRow.Kind.ISSUE, group)
                }
            }
        }
        return BackgroundContent("VIEWS", "${label(t.viewsTypeFilter)} · ${label(t.viewsDateFilter)}", label(t.viewsMode), rows)
    }

    private fun matchesDate(value: Instant?, filter: String): Boolean {
        if (filter == "ALL") return true
        if (value == null) return false
        val zone = ZoneId.systemDefault()
        val today = LocalDate.now(zone)
        val start = today.atStartOfDay(zone).toInstant()
        val tomorrow = today.plusDays(1).atStartOfDay(zone).toInstant()
        return when (filter) {
            "TODAY" -> value >= start && value < tomorrow
            "TOMORROW" -> value >= tomorrow && value < today.plusDays(2).atStartOfDay(zone).toInstant()
            else -> {
                val week = today.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY)).atStartOfDay(zone).toInstant()
                value >= week && value < week.plus(Duration.ofDays(7))
            }
        }
    }

    private fun formatDate(value: Instant): String = DATE_FORMAT.format(value.atZone(ZoneId.systemDefault()))
    private fun label(value: String) = value.lowercase().replace('_', ' ').replaceFirstChar { it.uppercase() }

    companion object { private val DATE_FORMAT = DateTimeFormatter.ofPattern("MMM d · HH:mm") }
}
